### ISSUE TITLE
cgroups hook issue getting systemd version

### ISSUE THREAD

**Issue by drtoss** [openpbs/openpbs#1818]
The cgroups hook routine _get_systemd_version invokes systemctl via subprocess.Popen() to learn the version of systemd by parsing the value out of stdout. Unfortunately, python3 changed Popen() to return the outputs as bytes arrays, rather than strings. You need to convert that to a string before using it in a regular expression test.

Suggested fix. Should work with python2, but untested.

```
--- /home7/dtalcott/work/drtaltc/src/hooks/cgroups/pbs_cgroups.PY       2020-06-08 12:02:08.930018109 -0700
+++ nas_cgroups.py      2020-06-09 08:52:47.835824332 -0700
@@ -3283,6 +3312,8 @@
                                        stdout=subprocess.PIPE,
                                        stderr=subprocess.PIPE)
             out, err = process.communicate()
+            if not isinstance(out, str):
+                out = out.decode('utf-8')
             ver = int(re.sub(r'Version=([0-9]+)', r'\1', out.splitlines()[0]))
         except Exception:
             # Note we also get here if we can't decode an int version

```

**Comment by alexis-cousein** [comment]
Could you post an example of an error generated by not converting to a str in Python3 in the MoM log (or evidence that the systemd version was not accurately detected)?

It would also be more readable to actually call "decode" only if the type is seen as bytes or bytearray (which are the types for which the method exists), rather than "not a string (so it must be a bytes or bytearray)".

**Comment by alexis-cousein** [comment]
An alternative would be to set universal_newlines to True in the Popen constructor. That seems even more elegant, given that we definitely don't mind universal newlines, and that should open everything in text mode and give us strings (not that having the guard behind it asa firewall is not still a good idea, mind you).

**Comment by drtoss** [comment]
> An alternative would be to set universal_newlines to True in the Popen constructor. That seems even more elegant, given that we definitely don't mind universal newlines, and that should open everything in text mode and give us strings in both Python 2.5-2.7 and Python 3 (having the guard to check the type behind it as a firewall is still a good idea, mind you).

That's what I have done in my own hooks. I wasn't sure which versions of python 2 supported universal newlines.

**Comment by alexis-cousein** [comment]
Indeed, that was broken:


    06/09/2020 13:48:37.102182;0800;pbs_python;Hook;pbs_python;_get_systemd_version: Method called 06/09/2020 13:48:37.161733;0800;pbs_python;Hook;pbs_python;__init__: systemd version seems to be 0

while on the host:

[cousein@pbi-33-r7 ~]$ systemctl --version 
systemd 208
+PAM +LIBWRAP +AUDIT +SELINUX +IMA +SYSVINIT +LIBCRYPTSETUP +GCRYPT +ACL +XZ

Unfortunately the test harness used to test this is itself in Python, so it usually makes the same error :-/.

Thanks for catching that. 

It's important for the long term health of jobs to correctly register pbs_jobs.service to systemd and ask for delegate control. 

It does work for current versions of PBSPro/OpenPBS with Python 2, but not for Python 3. 

The jobs on the Python 3 test systems were shortlived enough not to be harassed by systemd, so that was undetected...not that detecting interference from systemd is that easy, mind you, since it has a mind of its own.

**Comment by alexis-cousein** [comment]
Universal newlines were introduced in Python 2.3.

**Comment by alexis-cousein** [comment]
Not a good idea to check for "bytes" unless we know it exists. It doesn't in Python 2.5

**Comment by alexis-cousein** [comment]
OK, I see that there are many Popen calls that will suffer the same fate (e.g. if people crash through their memory limits and we execute dmesg to get the error message), so I think I'll just write a routine to sanity check the type (while always settting universal_newlines so that it's actually just going to return its argument).

**Comment by vstumpf** [comment]
Fixed in 6381a52
